contributors/design-proposals/service-injection-policy.md

```diff
@@ -0,0 +1,444 @@
+    * [Simple Pod Spec Example](#simple-pod-spec-example)
+    * [Pod Spec with `ConfigMap` Example](#pod-spec-with-`configmap`-example)
+    * [ReplicaSet with Pod Spec Example](#replicaset-with-pod-spec-example)
+
+
+## Abstract
+
+Describes a policy resource that allows loose coupling from services to the pods
+that consume them.
+
+## Motivation
+
+Consuming a service involves more than just connectivity.  In addition to
+coordinates to reach the service, credentials and non-secret configuration
+parameters are typically needed to use the service.  The primitives for this
+already exist, but a gap exists where loose coupling is desired: it should be
+possible to inject pods with the information they need to use a service on a
+service-by-service basis, without the pod authors having to incorporate the
+information into every pod spec where it is needed.
+
+## Constraints and Assumptions
+
+1.  Future work might require new mechanisms to be made to work with controllers
+    such as deployments and replicasets that create pods. Existing controllers
+    that create pods should recreate their pods when a new Service Injection
+    Policy is added that would effect them.
+
+## Use Cases
+
+1.  As a user, I want to be able to describe a way that pods should be injected
+    with the information to consume a particular service. This would allow me
+    to concisely model the information about how the service
+    should be consumed without altering every consuming pod spec.
+
+### Loose coupling between services and their consumers
+
+Using a Service Injection Policy allows pod template authors to not have to explicitly
+set information for every pod. This way authors of pod templates consuming a
+specific service do not need to know all the details about that service.
+
+The information will be injected based off the policy's `Selector`.
+
+## Proposed Changes
+
+### ServiceInjectionPolicy API object
+
+This resource is alpha. The policy itself is immutable.
+
+```go
+type ServiceInjectionPolicy struct {
+    unversioned.TypeMeta
+    ObjectMeta
+
+    Spec ServiceInjectionPolicySpec
+}
+
+type ServiceInjectionPolicySpec struct {
+    // Selector defines which Pods should be modified, this only targets Pods.
+    Selector      unversioned.LabelSelector
+    // Env defines the collection of EnvVar to inject.
+    Env           []EnvVar `json:omitempty`
+    // EnvFrom defines the collection of EnvFromSource to inject.
+    EnvFrom       []EnvFromSource `json:omitempty`
+    // Volumes defines the collection of Volume to inject.
+    Volumes       []Volume `json:omitempty`
+    // VolumeMounts defines the collection of VolumeMount to inject.
+    VolumeMounts  []VolumeMount `json:omitempty`
+}
+```
+
+#### Validations
+
+In order for the Service Injection Policy to be valid it must fulfill the
+following constraints:
+
+- The `Selector` field must be defined. This is how we know which pod templates
+  to inject so therefore it is required.
```

Comment:
I think we talked about how it would be nice to allow for a PIP to modify all pods - in that case should Selector be absent or will they be required to specify some kind of "all" selector clause?